From 99f00951a3c52fdc9df3dbcf5117cf365b3da032 Mon Sep 17 00:00:00 2001
From: "Uwe L. Korn" <uwe.korn@quantco.com>
Date: Mon, 23 Oct 2023 11:36:38 +0200
Subject: [PATCH] Use ConfigParser in versioneer

---
 versioneer.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/versioneer.py b/versioneer.py
index c010f63..4ddaee9 100644
--- a/versioneer.py
+++ b/versioneer.py
@@ -396,7 +396,7 @@ def get_config_from_root(root):
     # configparser.NoOptionError (if it lacks "VCS="). See the docstring at
     # the top of versioneer.py for instructions on writing your setup.cfg .
     setup_cfg = os.path.join(root, "setup.cfg")
-    parser = configparser.SafeConfigParser()
+    parser = configparser.ConfigParser()
     with open(setup_cfg, "r") as f:
         parser.readfp(f)
     VCS = parser.get("versioneer", "VCS")  # mandatory
-- 
2.42.0

